torch/nn/parallel/distributed.py

```diff
@@ -864,6 +866,49 @@ def __init__(
 
         self._lazy_init_ran = False
 
+        # Register the AccumulateGrad post hooks even if DDP is not compiled. This
+        # can avoid compiling the hooks twice. The Python hooks will be
+        # deregistered later if DDP is not compiled.
+        self._accum_grad_hooks: List[RemovableHandle] = []
+        self._ddp_python_hook = torch._dynamo.config.ddp_python_hook
+        self._force_to_disable_reducer = False
+        if self._ddp_python_hook:
+            if torch._dynamo.config.optimize_ddp:
+                raise RuntimeError(
+                    "Only one of `torch._dynamo.config.optimize_ddp` and "
+                    "`torch._dynamo.config.ddp_python_hook` should be True."
+                )
+            self._register_accum_grad_hook()
+
+    def _register_accum_grad_hook(self):
+        import torch.distributed._functional_collectives as fcol
+
+        def compiled_accum_grad_hook(
+            param,
+            *,
+            param_index: int,
+        ):
+            if not self.require_backward_grad_sync:
+                return
+
+            if self._comm_hooks:
```

Comment:
so in the case we have _comm_hooks, those would point to a fp16 or bf16 hook, which would do compression as well as grad scaling.  if no comm hooks, we do our own grad scaling?

Reply:
hmm, you added _comm_hooks as a new list in this PR, so it's not used by eager.

`dist._register_comm_hook(self.reducer, state, hook)` is used by eager.  Ideally i guess we could just peek at the list of hooks kept by reducer and use that directly.  but i guess that would require a new binding and that's annoying.  ok.

Reply:
although, a bit of a deviation, but, since we're deviating a bit anyway, maybe we should just create another comm hook that does the grad scaling, and register that hook in the case we do not register f/bf16 hook.  Then inside compiled_accum_grad_hook we can just call _comm_hooks unconditionally and no need to have a conditional/gradscaler code.

Reply:
Yes, we deviate it from the cpp reducer comm_hook because the original reducer comm_hook also have to obtain some cpp reducer information. Given that we disable cpp reducer, it makes sense to have a separate comm_hook list.